src/piadm.sh

```diff
@@ -195,6 +196,104 @@ DEFAULT_URL_PREFIX=https://us-central.manta.mnx.io/Joyent_Dev/public/SmartOS/
 # Default path for piadm's configuration
 PIADM_CONF=/var/piadm/piadm.conf
 
+# fetch_checksum
+#
+# Fetches a MD5 hash using a platform file as the key.
+#
+# Arguments:
+#   $1 - URL from where to fetch the PI from. The name of the PI
+#   	 is expected to be the last path component of the URL,
+#   	 or the literal string "latest".
+#
+# Environment Variables:
+#   PIADM_NO_CHECKSUM     - If set to 1, skips checksum validation.
+#   PIADM_CHECKSUM_URL    - If set, overrides the default md5sums.txt URL
+#   					  and uses the value provided.
+#
+# Notes:
+#   - If "latest" is used, up to 1024 PIs will be listed from ${URL_PREFIX}.
+#     The last one is assumed to be the latest.
+#   - On error, this function exits with code 1, following PIADM(8)
+#     convention: the error indicates a failure, but no changes were made
+#     to the system.
+checksum_platform=""
+fetch_checksum() {
+	if [[ -z "$checksum_platform" ]]; then
+		local platform_file
+		local stamp
+		if [ "$1" != "latest" ]; then
+			IFS=' ' read -r stamp platform_file <<<\
+			$(echo "$1" | rev | cut -d'/' -f1,2 | rev | tr -s "/" " ")
+		else
+			#latest is a special case, we need find out stamp
+			stamps_url="${URL_PREFIX%/}?limit=1024"
+			stamp=$("${CURL[@]}" "${stamps_url}" |\
+			json -ga -c "this.name.match(/Z$/)" | json -ag name |\
+				sort  | tail -1; exit ${PIPESTATUS[0]})
+			code=$?
+			if [[ $code -ne 0 ]]; then
+				eecho "Curl failed fetching PI data from ${stamps_url}"
+				return 1
+			fi
+			platform_file="smartos-${stamp}.iso"
+		fi
+		if [[ -z ${PIADM_CHECKSUM_URL} ]];then
+			local checksum_url="${URL_PREFIX}${stamp}/md5sums.txt"
+		else
+			local checksum_url="${PIADM_CHECKSUM_URL}"
+		fi
+		checksum_platform=$("${CURL[@]}"  "${checksum_url}" |\
+			awk -v pattern="${platform_file}"\
+			'$0 ~ pattern { print $1; exit}'; exit ${PIPESTATUS[0]})
+		code=$?
+		if [[ $code -ne 0 ]]; then
+			eecho "fetching checksums from ${checksum_url}"
+			return 1
+		fi
+		echo ${checksum_platform}
+		return 0
+	else
+		eecho "not recalculating checksum for $1" &>2
+		echo ${checksum_platform}
+		return 0
+	fi
+}
+
+# validate_checksum
+#
+# $1 is the URL from where the PI was downloaded.
+# $2 is the on disk PI image that was download from $1.
+#
+# Error code will always be 1, following the PIADM(8) convention where
+# a return code of 1 means: an error has occurred, but no change was made.
+# Environment variable PIADM_DIGEST_ALGORITHM controls the checksum 
+# algorithm used by DIGEST(1), by default md5 is used.  
+validate_checksum() {
+	if [[ $PIADM_NO_CHECKSUM -eq 1 ]]; then
+		vecho "WARNING: Not validation checksum"
+		return 0
+	fi
+	published_md5sum=$(fetch_checksum "$1")
+	code=$?
+	if [[ $code -ne 0 ]]; then
+		eecho "Could not get checksum for PI code: ${code}"
+		return 1
+	fi
+	local_md5sum=$(digest -a "${PIADM_DIGEST_ALGORITHM-md5}" "${2}")
```

Comment:
Yeah, especially since you can override it now.

Reply:
I'm tempted to include in this release as supplemental material `sha256sums.txt` to help you test this.

Reply:
(I'll have to hand-craft it for now, but eventually the jenkins build will do it automatically for at least one other algorithm.)

Reply:
See my one last new comment, but I'm closing this since hand-crafting one will be helpful.